In [1]:
import torch 
import torchvision 
from torchvision import transforms,datasets

In [2]:
train=datasets.MNIST('',train=True,download=True,transform=transforms.Compose([transforms.ToTensor()]))

In [3]:
test=datasets.MNIST('',train=False,download=True,transform=transforms.Compose([transforms.ToTensor()]))

In [4]:
trainset=torch.utils.data.DataLoader(train,batch_size=10,shuffle=True)
testset=torch.utils.data.DataLoader(test,batch_size=10,shuffle=False)

In [5]:
import torch.nn as nn 
import torch.nn.functional as F

### Its just like Object Oriented Programming 

✅ Building blocks to create and train neural networks easily

✅ Ready-made layers (like Linear, Conv2d), activation functions (like ReLU, Sigmoid), and loss functions (like CrossEntropyLoss)

✅ Tools to define models in a clean and modular way



In [10]:
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1=nn.Linear(28*28, 64)
        self.fc2=nn.Linear(64,64)
        self.fc3=nn.Linear(64,64)
        self.fc4=nn.Linear(64,10)

    def forward(self,x):
        x=F.relu(self.fx1(x))
        x=F.relu(self.fx2(x))
        x=F.relu(self.fx3(x))
        x=self.fx4(x)
        return F.log_softmax(x,dim=1)    
        
net=Net()
print(net)

Net(
  (fc1): Linear(in_features=784, out_features=64, bias=True)
  (fc2): Linear(in_features=64, out_features=64, bias=True)
  (fc3): Linear(in_features=64, out_features=64, bias=True)
  (fc4): Linear(in_features=64, out_features=10, bias=True)
)


✅ You are defining a new neural network class called Net.

✅ It inherits from nn.Module — which is the base class in PyTorch for all neural network modules.

By inheriting from nn.Module, your class gets lots of built-in features, like .parameters(), .to(), .eval(), etc.

---------------------------------------------------------------------------------

### Program Part

You are creating a fully connected(dense) layer called fc1.

nn.Linear is a layer from torch.nn that applies 
            output=input * weights+bias

28*28 → The number of input features = 784

This is common when working with MNIST images, which are 28×28 pixels.

The image is flattened into a 1D vector of length 784 before passing into the layer.



64 → The number of output features

The layer will produce a vector of length 64.

### This layer learns to transform the raw pixel data (size 784) into a smaller, 64-dimensional representation that is hopefully more useful for the task (like classification).